content/v2.0/actions/parameters.md

```diff
@@ -0,0 +1,308 @@
+Because `params[:book]` is `nil`, we can't access `:title`.
+In this case Ruby will raise a `NoMethodError`.
+
+We have a safe solution for our problem: `#get`.
+It accepts a list of symbols, where each symbol represents a level in our nested structure.
+
+```ruby
+params.get(:book, :title)             # => "Hanami"
+params.get(:unknown, :nested, :param) # => nil instead of NoMethodError
+```
+
+## Whitelisting
+
+In order to show how whitelisting works, let's create a new action: `app/actions/signup/create`
+
+We want to provide self-registration for our users.
+We build a HTML form which posts to an action that accepts the payload and stores it in the `users` table.
+That table has a boolean column `admin` to indicate whether a person has administration permissions.
+
+A malicious user can exploit this scenario by sending this extra parameter to our application, thereby making themselves an administrator.
+
+We can easily fix this problem by filtering the allowed parameters that are permitted inside our application.
+Please always remember that **params represent untrusted input**.
+
+We use `.params` to map the structure of the (nested) parameters.
+
+```ruby
+# app/actions/signup/create.rb
+
+module Bookshelf
+  module Actions
+    module Signup
+      class Create < Action
+        params do
+          required(:email).filled(:str?)
+          required(:password).filled(:str?)
+    
+          required(:address).schema do
+            required(:country).filled(:str?)
+          end
+        end
+    
+        def handle(req, res)
+          puts req.params[:email]             # => "alice@example.org"
+          puts req.params[:password]          # => "secret"
+          puts req.params[:address][:country] # => "Italy"
+    
+          puts req.params[:admin]             # => nil
+        end
+      end
+    end
+  end
+end
+```
+
+Even if `admin` is sent inside the body of the request, it isn't accessible from `params`.
+
+## Validations & Coercion
+
+### Use Cases
+
+In our example (called _"Signup"_), we want to make `password` a required param.
+
+Imagine we introduce a second feature: _"Invitations"_.
+An existing user can ask someone to join.
+Because the invitee will decide a password later on, we want to persist that `User` record without that value.
+
+If we put `password` validation in `User`, we need to handle these two use cases with a conditional.
+But in the long term this approach is painful from a maintenance perspective.
+
+```ruby
+# Example of poor style for validations
+class User
+  attribute :password, presence: { if: :password_required? }
+
+  private
+  
+  def password_required?
+     !invited_user? && !admin_password_reset?
+  end
+end
+```
+
+We can see validations as the set of rules for data correctness that we want for **a specific use case**.
+For us, a `User` can be persisted with or without a password, **depending on the workflow** and the route through
```

Comment:
```suggestion
For us, a `User` can be persisted with or without a password, **depending on the use case** and the route through
```

I think it's fine to repeat the same word here as `workflow` has specific meanings in various contexts (ie background job workflow etc.).